### PR TITLE
Fix `connect_timeout` for `wasi-http`

### PR DESCRIPTION
**Pull request by rikhuijzer**
@elliottt in https://github.com/bytecodealliance/wasmtime/issues/7356 mentions that `connect_timeout`, `first_byte_timeout`, and `between_bytes_timeout` should be tested.

This PR adds a test for `connect_timeout` and ensures that `TcpStream::connect` adheres to the timeout. Without this fix, setting `connect_timeout` has no effect on hosts where `TcpStream::connect` tries to connect.

Note that the variable `connect_timeout` is also used a bit later for the handshake:

https://github.com/bytecodealliance/wasmtime/blob/8dd2ae9e2c57b3fd32f91fb8f0ec1543854f4c7c/crates/wasi-http/src/types.rs#L207-L213

I'm not sure whether using this variable twice is correct.

### CODE REVIEW

File: crates/test-programs/src/bin/http_outbound_request_timeout.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    .context("/get");
+
+    assert!(res.is_err());
+    assert!(
+        matches!(
+            res.unwrap_err().downcast_ref::<ErrorCode>(),
+            Some(ErrorCode::ConnectionTimeout)
+        ),
+        "expected connection timeout"
+    );
```
Comment (elliottt) [inline crates/test-programs/src/bin/http_outbound_request_timeout.rs:32]: Thanks for matching the timeout explicitly here!

**Review by elliottt — APPROVED** [review]
This looks great, thank you!

**Comment by elliottt** [comment]
> I'm not sure whether using this variable twice is correct.

I see two uses in the `handler` function, but they're under different branches of the same conditional. Was there another use of `connect_timeout` that you were thinking of?

**Comment by rikhuijzer** [comment]
> This looks great, thank you!

Thanks for the review!

> > I'm not sure whether using this variable twice is correct.
> 
> I see two uses in the `handler` function, but they're under different branches of the same conditional. Was there another use of `connect_timeout` that you were thinking of?

There should be three (https://github.com/bytecodealliance/wasmtime/blob/eebce11eb1fdfc67ab8fc7a06cb68ba1231dd7a6/crates/wasi-http/src/types.rs).

One is now newly added near the `TcpStream::connect` and the other two are inside the branches near `hyper::client::conn::http1::handshake`. I think these two different phases represent TCP's SYN and ACK.

Why I mentioned it is because it sounded a bit weird to me that the `connect_timeout` is used twice, but on second thought maybe it's fine. The SYN and ACK are both still during the "Connection establishment" phase.

**Comment by elliottt** [comment]
Ah! Sorry for misunderstanding. I think this is fine to leave it this way, but I agree that it's a little odd. Reading through the handshake implementation, it's assuming that the connection has already been made, so if anything we might want to remove the timeout on the second two uses in the function.

**Comment by elliottt** [comment]
I added #8104 to record the fact that we should look into removing the uses of `connect_timeout` on the handshake calls, but that issue won't be valid until this PR merges.

**Pull request merged**